src/coreclr/jit/emitxarch.cpp

```diff
@@ -526,6 +526,35 @@ bool emitter::AreUpper32BitsZero(regNumber reg)
             }
         }
 
+        // This is a special case for cdq/cwde/cmpxchg.
+        if (instrHasImplicitRegSingleDest(id->idIns()))
+        {
+            switch (id->idIns())
+            {
+                case INS_cwde:
+                case INS_cmpxchg:
+                {
+                    if (reg == REG_RAX)
+                    {
+                        return PEEPHOLE_ABORT;
+                    }
+                    break;
+                }
+
+                case INS_cdq:
+                {
+                    if (reg == REG_RDX)
+                    {
+                        return PEEPHOLE_ABORT;
+                    }
+                    break;
+                }
+
+                default:
+                    break;
```

Comment:
Should we add unreachable/an assert here? Or maybe make `instrHasImplicitRegSingleDest` also return the written register to avoid duplicating the check for instructions?

Reply:
I was thinking this too. I'm going to bring in some of the work I was doing in the 'redundant cmp' for this PR.